### DISCUSSION TITLE
Testing PostFix (not receiving incoming emails)

### DISCUSSION THREAD

**Discussion by bernd-wechner** [modoboa/modoboa#2728] [Help]
I've installed a modoboa instance, and the website is working OK, and I have IMAP working fine as is the webmail interface. I used the standard installer, albeit fine-tuned the installer.cfg for my needs (I don't use nginx, and I am operating behind a reverse proxy).

I can use [https://mxtoolbox.com](https://mxtoolbox.com), and it is pretty happy with my SMTP server.  

Yet a test email I send myself from a Gmail account never arrives nor do I get a bounce. 

So I turned to an old workhorse in mail server testing, swaks which is very informative. The general test looks like:

```
$ swaks --port 587 -tls --server smtp-server.mydomain.tld --to me@mydomain.tld
```

and swaks essential reports PostFix rejecting it with:

```
<~* 554 5.7.1 <_gateway[192.168.0.1]>: Client host rejected: Access denied
```

What I lack is easy traction on why. The clue is in the message, namely that Postfix is getting the request via a reverse ptoxy on the gateway at 192.168.0.1 and I am fairly confident it relates to this config in `/etc/postfix/main.cf`:

```
# Recipient restriction rules
smtpd_recipient_restrictions =
      check_policy_service inet:127.0.0.1:9999
      permit_mynetworks
      permit_sasl_authenticated
      check_recipient_access
          proxy:pgsql:/etc/postfix/sql-maintain.cf
          proxy:pgsql:/etc/postfix/sql-relay-recipient-verification.cf
      reject_unverified_recipient
      reject_unauth_destination
      reject_non_fqdn_sender
      reject_non_fqdn_recipient
      reject_non_fqdn_helo_hostname
```

but I'm fairly green and there's a lot of reading in that and understanding it all, so I am hoping that someone here might save me some time and sport the issue. Maybe, Postfix needs to accept incoming connections and SMTP dialogues that it sees coming from the reverse proxy on 192.168.0.1. I'm not aware of any other (and I have many) services that even notice the reverse proxy. 

I simply have a port forward on the firewall there that sends all port 587 traffic to the server modoboa and postfix are on. The https and IMAP services for example run fine over that route.

**Comment by bernd-wechner** [comment]
Wow, finally worked this out. Took me ages but the reason was simply athentication!

```
$ swaks --port 587 -tls --server smtp-server.mydomain.tld --to me@mydomain.tld --auth-user me@mydomain.tld --auth-password mypassword
```

works!

Well, it now fails with `<~* 451 4.3.5 <bernd@thumbs.place>: Recipient address rejected: Server configuration problem` but that is another issue to resolve.

Turns out the issue was in facet in `master.cfg`:

```
submission inet n       -       -       -       -       smtpd
  -o syslog_name=postfix/submission
  -o smtpd_tls_security_level=encrypt
  -o tls_preempt_cipherlist=yes
  -o smtpd_sasl_auth_enable=yes
  -o smtpd_reject_unlisted_recipient=no
  -o smtpd_client_restrictions=permit_sasl_authenticated,reject
  -o smtpd_helo_restrictions=
  -o smtpd_sender_restrictions=reject_sender_login_mismatch
  -o milter_macro_daemon_name=ORIGINATING
  -o smtpd_proxy_filter=inet:[127.0.0.1]:10026
```

The clue being `  -o smtpd_client_restrictions=permit_sasl_authenticated,reject`.

Commenting that out and `swaks` worked. So put it back in again and authenticated with `swaks` and all good. So much time over such a small stupid oversight. As in doh! It might have helped if swaks (r postfix) reported an authorization error!

**Comment by bernd-wechner** [comment]
And so the saga continues and with swaks I'm making progress. swaks now reports:

```
<~* 451 4.5.0 Error in processing, id=188285-04, gen_mail_id FAILED: sql exec: err=7, 42501, DBD::Pg::st execute failed: ERROR:  permission denied for table maddr at (eval 123) line 173.
```

and this is accompanied in tthe postgreSQL logs with:

```
2023-01-02 06:00:41.624 UTC [2189195] amavis@modoboa-amavis ERROR:  permission denied for table maddr
2023-01-02 06:00:41.624 UTC [2189195] amavis@modoboa-amavis STATEMENT:  SELECT id FROM maddr WHERE partition_tag=$1 AND email=$2
```

I can seemingly confirm this with `psql`:

```
$ psql -d modoboa-amavis -U amavis -h localhost -c "select * from maddr"
ERROR:  permission denied for table maddr
$ psql -d modoboa-amavis -U modoboa -h localhost -c "select * from maddr"
ERROR:  permission denied for table maddr
```

I have an `amavis` and a `modoboa` user configured. And this is odd as `amavis` is the owner of the `modoboa-amavis` database which arises from my `installer.cfg`:
```
[modoboa]
user = modoboa
home_dir = /data/www/hermes
venv_path = /data/venv/modoboa
instance_path = %(home_dir)s
timezone = Australia/Hobart
dbname = modoboa
dbuser = modoboa
dbpassword = <password>
extensions = modoboa-amavis modoboa-pdfcredentials modoboa-postfix-autoreply modoboa-sievefilters modoboa-webmail modoboa-contacts modoboa-radicale
devmode = false

[amavis]
# a high-performance and reliable interface between mailer (MTA) and one or more content checkers: virus scanners,
# and/or Mail::SpamAssassin Perl module. It is written in Perl
enabled = true
user = amavis
max_servers = 2
dbname = modoboa-amavis
dbuser = amavis
dbpassword =  <password>
```

Anyhow that all went away when as the `postgres` user on the `modaboa-amavis` database I applied the sledge hammer of:

```
GRANT ALL PRIVILEGES ON DATABASE "modoboa-amavis" TO amavis;
GRANT ALL ON SCHEMA public TO amavis;
GRANT ALL PRIVILEGES ON ALL TABLES in SCHEMA public TO amavis;
GRANT ALL PRIVILEGES ON ALL SEQUENCES in SCHEMA public TO amavis;
```

The finally it progressed, but am now on an even more cryptic failure:

```
$ swaks --port 587 -tls --server smtp.myserver.tld --to me@myserver.tld --from me@myserver.tld --auth-user me@myserver.tld --auth-password <mypassword>
=== Trying smtp.myserver.tld:587...
=== Connected to smtp.myserver.tld.
<-  220 smpt.myserver.tld ESMTP
 -> EHLO myswaks
<-  250-smtp.myserver.tld
<-  250-PIPELINING
<-  250-SIZE 11534336
<-  250-ETRN
<-  250-STARTTLS
<-  250-ENHANCEDSTATUSCODES
<-  250-8BITMIME
<-  250-DSN
<-  250 CHUNKING
 -> STARTTLS
<-  220 2.0.0 Ready to start TLS
=== TLS started with cipher TLSv1.3:TLS_AES_256_GCM_SHA384:256
=== TLS no local certificate set
=== TLS peer DN="/CN=thumbs.place"
 ~> EHLO myswaks
<~  250-smtp.myserver.tld
<~  250-PIPELINING
<~  250-SIZE 11534336
<~  250-ETRN
<~  250-AUTH PLAIN LOGIN
<~  250-AUTH=PLAIN LOGIN
<~  250-ENHANCEDSTATUSCODES
<~  250-8BITMIME
<~  250-DSN
<~  250 CHUNKING
 ~> AUTH LOGIN
<~  334 base64stuff
 ~> base64stuff
<~  334 base64stuff
 ~> base64stuff
<~  235 2.7.0 Authentication successful
 ~> MAIL FROM:<me@myserver.tld>
<~  250 2.1.0 Ok
 ~> RCPT TO:<me@myserver.tld>
<~  250 2.1.5 Ok
 ~> DATA
<~  354 End data with <CR><LF>.<CR><LF>
 ~> Date: Mon, 02 Jan 2023 17:27:50 +1100
 ~> To: me@myserver.tld
 ~> From: me@myserver.tld
 ~> Subject: test Mon, 02 Jan 2023 17:27:50 +1100
 ~> Message-Id: <20230102172750.194459@myswaks>
 ~> X-Mailer: swaks v20190914.0 jetmore.org/john/code/swaks/
 ~> 
 ~> This is a test mailing
 ~> 
 ~> 
 ~> .
<~* Timeout (30 secs) waiting for server response
 ~> QUIT
<~* Timeout (30 secs) waiting for server response
=== Connection closed with remote host.
```

Hmmm... the trials.

**Comment by bernd-wechner** [comment]
And the clues lie in `/var/mail/log` watched while testing with `swaks` reveals the crucial:

```
Jan  2 17:56:22 myserver amavis[188286]: (188286-06) (!)connect to /var/run/clamav/clamd.ctl failed, attempt #1: Can't connect to a UNIX socket /var/run/clamav/clamd.ctl: No such file or directory
Jan  2 17:56:23 myserver amavis[188286]: (188286-06) (!)connect to /var/run/clamav/clamd.ctl failed, attempt #1: Can't connect to a UNIX socket /var/run/clamav/clamd.ctl: No such file or directory
Jan  2 17:56:23 myserver amavis[188286]: (188286-06) (!)ClamAV-clamd: All attempts (1) failed connecting to /var/run/clamav/clamd.ctl, retrying (2)
Jan  2 17:56:29 myserver amavis[188286]: (188286-06) (!)connect to /var/run/clamav/clamd.ctl failed, attempt #1: Can't connect to a UNIX socket /var/run/clamav/clamd.ctl: No such file or directory
Jan  2 17:56:29 myserver amavis[188286]: (188286-06) (!)ClamAV-clamd av-scanner FAILED: run_av error: Too many retries to talk to /var/run/clamav/clamd.ctl (All attempts (1) failed connecting to /var/run/clamav/clamd.ctl) at (eval 128) line 659.\n
Jan  2 17:56:29 myserver amavis[188286]: (188286-06) (!)WARN: all primary virus scanners failed, considering backups
```

The sleuth now turns to Clam!

In `installer.cfg` I had:

```
[clamav]
# open-source standard for mail gateway-scanning software
enabled = true
user = clamav
````

**Reply by bernd-wechner** [reply]
Sorted this one by running `sudo dpkg-reconfigure clamav-daemon` and walking through that then `sudo service clamav-daemon restart` and all came good.

**Comment by bernd-wechner** [comment]
Having sorted ClamAV the issue now is OpenDKIM:

```
Jan  2 18:47:20 myserver postfix/smtpd[2220640]: connect from localhost[127.0.0.1]
Jan  2 18:47:20 myserver postfix/smtpd[2220640]: D2F121520732: client=localhost[127.0.0.1], orig_client=_gateway[192.168.0.1]
Jan  2 18:47:20 myserver postfix/cleanup[2219701]: D2F121520732: message-id=<20230102184717.211519@mysource>
Jan  2 18:47:20 myserver opendkim[2561]: KeyTable entry for '1' corrupt
Jan  2 18:47:20 myserver opendkim[2561]: D2F121520732: error loading key '1'
Jan  2 18:47:20 myserver postfix/cleanup[2219701]: D2F121520732: milter-reject: END-OF-MESSAGE from localhost[127.0.0.1]: 4.7.1 Service unavailable - try again later; from=<me@myserver.tld> to=<me@myserver.tld> proto=ESMTP helo=<mysource>
Jan  2 18:47:20 myserver postfix/smtpd[2220640]: disconnect from localhost[127.0.0.1] ehlo=1 xforward=1 mail=1 rcpt=1 data=0/1 quit=1 commands=5/6
```

The trials and tribulations.

This might be related:

```
$ sudo service opendkim status
● opendkim.service - OpenDKIM Milter
     Loaded: loaded (/lib/systemd/system/opendkim.service; enabled; vendor preset: enabled)
     Active: active (running) since Thu 2022-12-29 14:38:29 AEDT; 4 days ago
       Docs: man:opendkim(8)
             man:opendkim.conf(5)
             man:opendkim-lua(3)
             man:opendkim-genkey(8)
             man:opendkim-genzone(8)
             man:opendkim-testkey(8)
             http://www.opendkim.org/docs.html
   Main PID: 2561 (opendkim)
      Tasks: 6 (limit: 6986)
     Memory: 4.2M
        CPU: 14min 49.826s
     CGroup: /system.slice/opendkim.service
             └─2561 /usr/sbin/opendkim

Jan 02 19:16:44 myserver opendkim[2561]: C171215207CD: no signing table match for 'modoboa@smtp.myserver.tld'
Jan 02 19:16:44 myserver opendkim[2561]: C171215207CD: no signature data
Jan 02 19:16:45 myserver opendkim[2561]: 0EF0915207CD: no signing table match for 'modoboa@smtp.myserver.tld'
Jan 02 19:16:45 myserver opendkim[2561]: 0EF0915207CD: no signature data
Jan 02 19:16:45 myserver opendkim[2561]: 4EC4415207CD: no signing table match for 'modoboa@smtp.myserver.tld'
Jan 02 19:16:45 myserver opendkim[2561]: 4EC4415207CD: no signature data
Jan 02 19:16:45 myserver opendkim[2561]: 8E52615207CD: no signing table match for 'modoboa@smtp.myserver.tld'
Jan 02 19:16:45 myserver opendkim[2561]: 8E52615207CD: no signature data
Jan 02 19:16:45 myserver opendkim[2561]: CFFAA15207CD: no signing table match for 'modoboa@smtp.myserver.tld'
Jan 02 19:16:45 myserver opendkim[2561]: CFFAA15207CD: no signature data
```

And in the database `select * from dkim;` reveals only one entry wither:

id: 1
domain_name: myserver.tld
private_key_path: <empty>
selector: modoboa


The KeyTable definition is:

```
KeyTable		dsn:pgsql://opendkim:password@5432+127.0.0.1/modoboa/table=dkim?keycol=id?datacol=domain_name,selector,private_key_path
```

which is visible in in PostgreSQL as:

`SELECT domain_name,selector,private_key_path FROM dkim WHERE id = '1'`

and I might rightly suspect openDKIM is looking up modoboa@smtp.myserver.tld and getting back selector@domain_name as modoboa@myserver.tld.

Nose down. Thus far unable to find any docs on openDKIM telling use what KeyTable expects. Watch this space.

Nor can I see where in the openDKIM configs it even gets modoboa@smtp.myserver.tld.

**Reply by bernd-wechner** [reply]
OK, I suspect the `modoboa` in `modoboa@smtp.myserver.tld` comes from the DKIM selector I specified:

![image](https://user-images.githubusercontent.com/7296506/210219601-a45ea67f-bc58-4ae8-9bc9-a11543b24dcd.png)

though I'm still a tad stuck on whence it gets the `smtp.myserver.tld`. Further I can try it manually:

```
$ /usr/sbin/opendkim -vv
opendkim: using default configfile /etc/opendkim.conf
opendkim: /etc/opendkim.conf: could not find valid key record "1" in KeyTable
```
and see a similar message with a useless message (https://github.com/trusteddomainproject/OpenDKIM/issues/164) though openDKIM looks like abandonware alas leaving us to plod our way through using it with the code and doc as they are. 

But it gets worse, I restarted it, and it wouldn't even start:

```
$ sudo service opendkim restart
Job for opendkim.service failed because the control process exited with error code.
See "systemctl status opendkim.service" and "journalctl -xeu opendkim.service" for details.
$ service opendkim status
× opendkim.service - OpenDKIM Milter
     Loaded: loaded (/lib/systemd/system/opendkim.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Mon 2023-01-02 21:32:27 AEDT; 3min 24s ago
       Docs: man:opendkim(8)
             man:opendkim.conf(5)
             man:opendkim-lua(3)
             man:opendkim-genkey(8)
             man:opendkim-genzone(8)
             man:opendkim-testkey(8)
             http://www.opendkim.org/docs.html
    Process: 2271257 ExecStart=/usr/sbin/opendkim (code=exited, status=78)
        CPU: 33ms

Jan 02 21:32:27 arachne systemd[1]: opendkim.service: Scheduled restart job, restart counter is at 5.
Jan 02 21:32:27 arachne systemd[1]: Stopped OpenDKIM Milter.
Jan 02 21:32:27 arachne systemd[1]: opendkim.service: Start request repeated too quickly.
Jan 02 21:32:27 arachne systemd[1]: opendkim.service: Failed with result 'exit-code'.
Jan 02 21:32:27 arachne systemd[1]: Failed to start OpenDKIM Milter.
$ cat /lib/systemd/system/opendkim.service
[Unit]
Description=OpenDKIM Milter
Documentation=man:opendkim(8) man:opendkim.conf(5) man:opendkim-lua(3) man:opendkim-genkey(8) man:opendkim-genzone(8) man:opendkim-testkey(8) http://www.opendkim.org/docs.html
After=network.target nss-lookup.target postgresql.service postgresql.service postgresql.service

[Service]
Type=forking
PIDFile=/run/opendkim/opendkim.pid
ExecStart=/usr/sbin/opendkim
ExecReload=/bin/kill -USR1 $MAINPID
Restart=on-failure

[Install]
WantedBy=multi-user.target
```
and from `journalctl`:

```
Jan 02 21:32:27 arachne systemd[1]: opendkim.service: Failed with result 'exit-code'.
     Subject: Unit failed
	 Defined-By: systemd
	 Support: http://www.ubuntu.com/support
	 
	 The unit opendkim.service has entered the 'failed' state with result 'exit-code'.
Jan 02 21:32:27 arachne systemd[1]: Failed to start OpenDKIM Milter.
	 Subject: A start job for unit opendkim.service has failed
	 Defined-By: systemd
	 Support: http://www.ubuntu.com/support
	 
	 A start job for unit opendkim.service has finished with a failure.
	 
	 The job identifier is 17027 and the job result is failed.
```

Oh, the lucid and useful messages ... and all systemd is doing is starting `/usr/sbin/opendkim`

On another frontm the reason it's unhappy is probably that the private_key_path is missing. and so one area toe xpore is:

` python manage.py modo manage_dkim_keys`

which I'll do and report back.

**Comment by PatTheMav** [comment]
First things first: Did you use the installer or did you build this setup yourself?

As to your journey of discovery:

`smtpd_recipient_restrictions` sets a default set of restrictions - it is good to be very strict here and use the option overrides (`-o`) in the `master.cf` file to customise restrictions for the specific services. Passing an empty value for a specific restriction effectively disables it.
* Modern postfix uses a sane default set of restrictions for port 25: It only accepts emails for which it is the final destination (disabling relay) and for existing mailboxes

In your case you tried to use the `submission` port (587) to send an email via your `postfix` instance to another server (it "relays" the mail for you) - because we don't want open relays, additional restrictions apply (as configured in `master.cf`):

* The `AUTH` extension is only enabled for port 587 (`smtpd_sasl_auth_enable=yes`)
* Encryption via `STARTTLS` is required and not optional (`smtpd_tls_security_level=encrypt`)
* Clients that are not authenticated via `SASL` are rejected (`smtpd_client_restrictions=permit_sasl_authenticated,reject`)

Next comes `amavis` - the current modoboa instance uses `amavis` as a proxy for SpamAssassin and ClamAV checks:

* Mail is offhanded to `amavis` by `postfix` on port 10024
* `amavis` runs its checks, adds headers
* `amavis` sends the email again, only this time via port 10025, where an `smtpd` instance with almost no restrictions accepts mails (which is also why you should never allow that port to be available to the world)

`amavis` itself keeps spam analysis data in the database set up as part of modoboa:

* _If_ installed correctly, you should have two separate database users `amavis` and `spamassassin`, both with respective databases of the same name
* The `amavis` user by default gets full access to the `modoboa` database: `GRANT ALL ON DATABASE amavis TO modoboa`

As for `OpenDKIM`, it uses a predefined view created in the modoboa database that it needs to have access to: `GRANT SELECT ON dkim TO opendkim`. This table should usually yield the following fields:
* `id` - numerical id of the entry
* `domain_name` - the domain name used to send emails from, needs to have the public DKIM key set up in its DNS setup
* `private_key_path` - absolute path to the private key files generated by modoboa
* `selector` - the DNS selector used for DKIM public keys, which need to be set up as a TXT entry available via `<SELECTOR>._domainkey.<YOURDOMAIN>`

So when signing outgoing mails, OpenDKIM will look up the domain name used in the mail, look up its ID via the `SigningTable` configuration parameter, then look up the private key path using the `KeyTable` with the ID yielded of the first query.
* It seems your `private_key_path` is empty, meaning that modoboa did not generate the necessary DKIM public+private keys yet and update the database entry
* You need to create a public and private key pair and add the public key to your DNS (the entry is helpfully provided by the modoboa admin UI)

---

TL;DR: Did you use the installer? Did it finish without errors? These issues should not occur if the system was properly installed or _if_ you installed it manually you knew what you were doing 🙃...

**Reply by bernd-wechner** [reply]
@PatTheMav, thanks enormously for sharing such an informative insight into the process. Much of this I have already learned during the process to date but far from all of it. Notably, the role of amavis and port 10025 is news to me (and welcome).

Other thoughts:

1. I used the installer, albeit I edited the installer.cfg rather more heavily than most might (disabled nginx install primarily but walked through it and generally reviewed)
2. I have port 25 closed on the firewall. The world is going entirely SSL based rapidly and part of me is surprised 25 is even still on the radar. Indeed all my mail services run over STARTLS. `swaks` is an awesome toolkit and comms with postfix are fine (since I kicked myself for forgetting to use AUTH with `swaks` and failing to interpret the first hurdle (`Client host rejected: Access denied` message from postfix as a simple "authorisation failed" message).
3. I'll get onto DKIM soon ... thanks deeply for the confirmation that private_key is needed (all part of the learning here) and of course I'll double check I have a TXT entry set up (I do but it's probably not right, as I'm trying to migrate from one service to self hosted).

TLDR: Yep used the installer, no errors, In fact I keep extensive notes and here is the trace (redacted - I tend to remove identifying info from public forum posts to lower the bot harvest rewards):

```
$ sudo ./run.py --interactive myserver.tld
Welcome to Modoboa installer!

Warning:
Before you start the installation, please make sure the following DNS records exist for domain 'myserver.tld':
  smpt IN A   <IP ADDRESS OF YOUR SERVER>
       IN MX  smtp.myserver.tld.

Your mail server will be installed with the following components:
modoboa automx amavis clamav dovecot razor postfix postwhite spamassassin radicale opendkim backup
Do you confirm? (Y/n) Y
The process can be long, feel free to take a coffee and come back later ;)
Starting...
Installing amavis
Installing spamassassin
Installing razor
Installing clamav
Installing automx
User automx already exists, skipping creation but please make sure the /srv/automx directory exists.
Installing radicale
User radicale already exists, skipping creation but please make sure the /srv/radicale directory exists.
Installing opendkim
Installing postfix
Installing postwhite
Installing dovecot
User dovecot already exists, skipping creation but please make sure the /srv/vmail directory exists.
Congratulations! You can enjoy Modoboa at https://smtp.myserver.tld (admin:password)
```

And yes, precisely one of the draws to modoboa! All in one simple install. Alas, I tripped over a few hurdles. No crisis, I actually value the learning. But it is a tad time-consuming alas.

**Reply by PatTheMav** [reply]
You'd be surprised how often people fall into this trap: Do **not** close port 25!

Email is not like other protocols in that it uses fully-encrypted transport - current RFC guidance requires you to use opportunistic encryption:

* Accept incoming mail server connections on port 25, encrypt _if_ possible `smtpd_tls_security_level=may`.
* This will allow clients that _do_ support TLS encryption to upgrade the connection via the `STARTTLS` command
* Allow falling back to unencrypted mail delivery otherwise

This does not apply to submission on port 587, where you can require your authenticated mail clients to transmit and authenticate only via a secure connection (which is the default).

On top of that, other email servers will not talk to your server on any other port but port 25. It is the default `smtp` port for a reason.

**Reply by bernd-wechner** [reply]
Will get onto that once I have `swaks` testing complete. `swaks` rocks! ;-)

**Comment by bernd-wechner** [comment]
And the DKIM issues are resolved.

In a nutshell we need to run:

`python manage.py modo manage_dkim_keys`

and it needs to work (I filed https://github.com/modoboa/modoboa/issues/2738 and https://github.com/modoboa/modoboa/issues/2737 in the process of getting it to work) and drop a `.pem` file or files for you domain or domains in the configured `DKIM keys storage directory` that can be found at `https://your.modoboa.site/core/#parameters/`. 

Once done and the `.pem` file or files are there then voila:

```
$ sudo service opendkim restart
(modoboa) weaver@arachne:/data/www/hermes$ sudo service opendkim status
● opendkim.service - OpenDKIM Milter
     Loaded: loaded (/lib/systemd/system/opendkim.service; enabled; vendor preset: enabled)
    Drop-In: /etc/systemd/system/opendkim.service.d
             └─override.conf
     Active: active (running) since Tue 2023-01-03 17:18:36 AEDT; 1s ago
       Docs: man:opendkim(8)
             man:opendkim.conf(5)
             man:opendkim-lua(3)
             man:opendkim-genkey(8)
             man:opendkim-genzone(8)
             man:opendkim-testkey(8)
             http://www.opendkim.org/docs.html
    Process: 2594867 ExecStart=/usr/sbin/opendkim -P /run/opendkim/opendkim.pid -p inet:12345@localhost (code=exited, status=0/SUCCESS)
   Main PID: 2594878 (opendkim)
      Tasks: 6 (limit: 6986)
     Memory: 5.8M
        CPU: 61ms
     CGroup: /system.slice/opendkim.service
             └─2594878 /usr/sbin/opendkim -P /run/opendkim/opendkim.pid -p inet:12345@localhost
```

and finally, a delivered mail:

```
$ swaks --port 587 -tls --server smtp.myserver.tld --to me@myserver.tld --from me@myserver.tld --auth-user me@myserver.tld --auth-password mypassword
=== Trying smtp.myserver.tld:587...
=== Connected to smtp.myserver.tld.
<-  220 smtp.myserver.tld ESMTP
 -> EHLO swakshost
<-  250-smtp.myserver.tld
<-  250-PIPELINING
<-  250-SIZE 11534336
<-  250-ETRN
<-  250-STARTTLS
<-  250-ENHANCEDSTATUSCODES
<-  250-8BITMIME
<-  250-DSN
<-  250 CHUNKING
 -> STARTTLS
<-  220 2.0.0 Ready to start TLS
=== TLS started with cipher TLSv1.3:TLS_AES_256_GCM_SHA384:256
=== TLS no local certificate set
=== TLS peer DN="/CN=myserver.tld"
 ~> EHLO swakshost
<~  250-smtp.myserver.tld
<~  250-PIPELINING
<~  250-SIZE 11534336
<~  250-ETRN
<~  250-AUTH PLAIN LOGIN
<~  250-AUTH=PLAIN LOGIN
<~  250-ENHANCEDSTATUSCODES
<~  250-8BITMIME
<~  250-DSN
<~  250 CHUNKING
 ~> AUTH LOGIN
<~  334 base64fluff
<~  235 2.7.0 Authentication successful
 ~> MAIL FROM:<me@myserver.tld>
<~  250 2.1.0 Ok
 ~> RCPT TO:<me@myserver.tld>
<~  250 2.1.5 Ok
 ~> DATA
<~  354 End data with <CR><LF>.<CR><LF>
 ~> Date: Tue, 03 Jan 2023 17:06:25 +1100
 ~> To: me@myserver.tld
 ~> From: me@myserver.tld
 ~> Subject: test Tue, 03 Jan 2023 17:06:25 +1100
 ~> Message-Id: <20230103170625.617444@swakshost>
 ~> X-Mailer: swaks v20190914.0 jetmore.org/john/code/swaks/
 ~> 
 ~> This is a test mailing
 ~> 
 ~> 
 ~> .
<~  250 2.0.0 from MTA(smtp:[127.0.0.1]:10025): 250 2.0.0 Ok: queued as 1B79815205ED
 ~> QUIT
<~  221 2.0.0 Bye
=== Connection closed with remote host.
```

**Reply by PatTheMav** [reply]
That call should usually run automatically as a cron job, so there is no need to run it manually.

**Reply by bernd-wechner** [reply]
The installer did not install any cron jobs to that effect. In fact, the only reason I know about them is from the manual install instructions:

https://modoboa.readthedocs.io/en/latest/manual_installation/modoboa.html#cron-jobs

As an aside given the ubiquity (love it or not) of systemd and that modoboa installs numerous systemd based packages like dovecot, postfix, amavis, opendkim, I have always wondered why cron is still used for scheduling jobs and not systemd timers. Just a curio regarding consistency really (and having used cron now for about 35 years I still find its crontab syntax a dog's breakfast ;-)

**Reply by PatTheMav** [reply]
Yeah just a tip - familiarise yourself with all the things the installer does, the custom files it installs and all, so you don't continuously run into these issues (I've manually set up and heavily customised my modoboa installation a few years back and looking at the installer in-depth answered lots of questions by itself).

Having customised the installer script, you are mostly on your own because the fine folks here couldn't possibly support customisations on top of the existing scope of what the installer does.

The crontab and other support files are defined in the default script (https://github.com/modoboa/modoboa-installer/blob/5c22600d98a19cedc7ce385c93e5ed998c639e6e/modoboa_installer/scripts/modoboa.py#L37) so they should've been configured. If that part is missing, many other important parts might be as well, so trace every step the installer would have made and cross-check if it was actually set up for you.

As for the timers, I'd guess it's because lots of people are unfamiliar with the full scope of `systemd`'s capabilities (they know about cron, but might not know about timers) and a still ongoing hostility towards its in the Linux world.

**Reply by bernd-wechner** [reply]
Egads, can you kindly save me from pulling my hair out over opendkim. This package is batshit crazy IMHO. Sorry. 

I just spent ages finding user:group and file permission to keep the danged service quiet. It's complaining endlessly in log file.

But only about the second domain I added. The first one no complaint. The second one a litany of them.

First it's: `2: can't load key from /etc/mail/DKIMkeys/mydomain2.tld.pem: Permission denied`

So I check the process and evolve a one-liner to get the info:

```
$ ps -o uname,ruser,group,rgroup,supgrp -p `systemctl show opendkim.service --property=MainPID | cut -d = -f 2`
USER     RUSER    GROUP    RGROUP   SUPGRP
opendkim opendkim opendkim opendkim opendkim
```
so I set the perms on them as: `-rw-r----- 1 root opendkim`

Thinking for security only root and opendkim can read it and root write it, so when generating them I use `sudo`. They were root:root.

Then it moans in bright orange: `2: key data is not secure: opendkim is in group 8 which has multiple users (e.g., "mail")`

But no, not at all. I installed the `members` package for neatness (because greps on /etc/group suck, and always leave a shadow of doubt IMHO) and:

```
$ members opendkim
opendkim
```

But it does mention group 8 so:

```
$ grep :8: /etc/group
mail:x:8:modoboa
```

But how is it relevant my pem files:

```
$ ll /etc/mail/DKIMkeys
total 16
drwxrwxr-x 2 root root 4096 Jan  4 22:09 ./
drwxr-xr-x 4 root root 4096 Dec 29 18:23 ../
-rw-r----- 1 root opendkim 1704 Jan  4 22:09 mydomain.tld.pem
-rw-r----- 1 root opendkim 1704 Jan  3 17:05 mydomain2.tld.pem
```

What has group got to do with anything?  Though it does have a second user 'modoboa', I'll grant it that.

So I found a setting recommended to get rid of this message: `RequireSafeKeys false`. I added it to the conf and restarted. Wow, great feature, now it floods the logs with the same danged message but in yellow not deep orange.

Right, on a whim I throw a sledge hammer at it, and  set them up like:

```
$ ll /etc/mail/DKIMkeys
total 16
drwxrwxr-x 2 root root 4096 Jan  4 22:09 ./
drwxr-xr-x 4 root root 4096 Dec 29 18:23 ../
-rw------- 1 opendkim opendkim 1704 Jan  4 22:09 mydomain.tld.pem
-rw------- 1 opendkim opendkim 1704 Jan  3 17:05 mydomain2.tld.pem
```

And it's finally happy! Stopped flooding the log with those errors.

So quick question, what ownerships and permissions are you using?  And if I add a domain, my guess is this won't work with:

`$ python manage.py modo manage_dkim_keys`

without some special handling. Even if that runs under cron, it either run as root to write the keys (currently needed a only root can write to the directory) and then the new key is owned by root, or it runs as the opendkim user and I fix the permissions to:

```
$ ll /etc/mail/DKIMkeys
total 16
drwxrwx--- 2 root opendkim 4096 Jan  4 22:09 ./
drwxr-xr-x 4 root root 4096 Dec 29 18:23 ../
-rw------- 1 opendkim opendkim 1704 Jan  4 22:09 mydomain.tld.pem
-rw------- 1 opendkim opendkim 1704 Jan  3 17:05 mydomain2.tld.pem
```
That would work on concert. But as noted, curious how you're set up?

As an aside have port 25 open and working, but am still working on delivery, works with swaks, not from yahoo.com yet. Was trying first remove a flood of these openkim errors in the logs so I can watch a mail log while I send from yahoo.

**Reply by PatTheMav** [reply]
From the OpenDKIM readme (http://www.opendkim.org/staging/opendkim-README):

> (4) Store the private key in a safe place.  We generally use a path like
    /var/db/dkim/SELECTOR.key.pem (where "SELECTOR" is the name you chose).
    The /var/db/dkim directory and the associated .pem file should be owned by
    the user that will be executing the filter (preferably not the
    superuser) and be mode 0700 and 0600 respectively.

In modoboa's case, the default directory is `/var/lib/dkim` and should be owned by `opendkim`, as you can tell by the installer script:

https://github.com/modoboa/modoboa-installer/blob/5c22600d98a19cedc7ce385c93e5ed998c639e6e/modoboa_installer/scripts/opendkim.py#L35-L50

Once again - please familiarise yourself with what the installer does and also refer to the documentation of the tools used by modoboa, as both do answer most of your questions you had so far.

**Reply by PatTheMav** [reply]
This goes hand-in-hand with the generation of the keys, as you can see from the `crontab` template, `manage_dkim_keys` needs to be executed by the opendkim user to ensure access:

https://github.com/modoboa/modoboa-installer/blob/5c22600d98a19cedc7ce385c93e5ed998c639e6e/modoboa_installer/scripts/files/modoboa/crontab.tpl#L36

**Reply by bernd-wechner** [reply]
Thanks again ever so much for your patience and kind support. Am, working on familiarisation.  It's a slow process and not always easy (well, not generally easy) to find where these things are done in the installer. But the crontab is a gem insight, didn't notice it specified the use. You're a champ @PatTheMav.
